espnet2/spk/espnet_model.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2023 Jee-weon Jung
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+from contextlib import contextmanager
+from itertools import permutations
+from typing import Dict, Optional, Tuple
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from packaging.version import parse as V
+from torch.cuda.amp import autocast
+from typeguard import check_argument_types
+
+# TODO: remove unused modules
+from espnet2.asr.encoder.abs_encoder import AbsEncoder
+from espnet2.asr.frontend.abs_frontend import AbsFrontend
+from espnet2.asr.specaug.abs_specaug import AbsSpecAug
+from espnet2.layers.abs_normalize import AbsNormalize
+from espnet2.spk.encoder.rawnet3_encoder import RawNet3Encoder
+from espnet2.spk.loss import *
+from espnet2.spk.pooling.abs_pooling import AbsPooling
+from espnet2.spk.projector.abs_projector import AbsProjector
+from espnet2.torch_utils.device_funcs import force_gatherable
+from espnet2.train.abs_espnet_model import AbsESPnetModel
+from espnet.nets.pytorch_backend.nets_utils import to_device
+
+
+class ESPnetSpeakerModel(AbsESPnetModel):
+    """
+    Speaker embedding extraction model.
+    Core model for diverse speaker-related tasks (e.g., verification, open-set
+    identification, diarization)
```

Comment:
How about adding today's discussions of a unique structure of using `pooling` and `projector` modules here? It would be good to describe them here.

Reply:
We could also add some discussions of possible replacement of the `decoder` module in the future.